### PR TITLE
(PUP-10653) Remove Dir monkey patch

### PR DESCRIPTION
**Pull request by GabrielNagy**
We removed our external dependency on the win32-dir gem in https://github.com/puppetlabs/puppet/pull/8258 by pulling the gem code in puppet. However, it seems that win32-dir features are not used outside of constants which can all be replaced by environment variables. Do that, and remove the monkey patched code.

### CODE REVIEW

**Comment by GabrielNagy** [comment]
Since almost all the places that access the constant also require `win32/dir`, I'd be more inclined to update the modules to use the environment variable. Even with the monkeypatch present the require statement would still fail as we no longer vendor the gem.

Bolt as a gem will probably need to have an explicit dependency on the gem, while packaged Bolt will still work as their runtime includes the win32 gems: https://github.com/puppetlabs/puppet-runtime/blob/26a3d6eaf8f1ef750e73d785583a60c48a7495ec/configs/projects/bolt-runtime.rb#L197.

I think it's gnarlier if we keep the monkey patch and another project requires `win32/dir` on top of it as they'd get warnings due to redefined constants (which can be mitigated on our side, but still). It seems that either way we choose to do it, those places will need to be updated.

**Comment by GabrielNagy** [comment]
Ok, I went through all the places where we call `glob` and `[]` and we don't appear to call the methods with backslash-separated paths. I annotated all occurences here https://gist.github.com/GabrielNagy/bbe659d8268d924d668f91b3f05a7b76, nothing looks problematic.

`Dir.pwd` is used here: https://github.com/puppetlabs/puppet/blob/main/lib/puppet/application/resource.rb#L176
Calling `#path` on this produces a mix of forward and backslashes if `win32/dir` is loaded, but this is used only when running `puppet resource` with `--edit` which doesn't seem to work on Windows. So at best, it would be an improvement if we remove the monkey patched `Dir.pwd`.

About `SmartPath`, `start_with?` is called with `generic_path` which is always something that goes through `File.join` which uses forward slashes.

6767686 and aa67f14 don't cause any breakage after removing the monkey patched code, but I'd still revert them to avoid eventual confusion.  Thoughts @joshcooper?

File: ext/windows/service/daemon.rb [hunk 1]
```diff
@@ -11,16 +11,13 @@
...
-require 'puppet/util/windows/monkey_patches/dir'
-
 class WindowsDaemon < Puppet::Util::Windows::Daemon
   CREATE_NEW_CONSOLE          = 0x00000010
 
   @run_thread = nil
   @LOG_TO_FILE = false
   @loglevel = 0
-  LOG_FILE =  File.expand_path(File.join(Dir::COMMON_APPDATA, 'PuppetLabs', 'puppet', 'var', 'log', 'windows.log'))
+  LOG_FILE =  File.expand_path(File.join(ENV['ALLUSERSPROFILE'], 'PuppetLabs', 'puppet', 'var', 'log', 'windows.log'))
```
Comment (joshcooper) [inline ext/windows/service/daemon.rb:20]: If the service is configured to run under a domain service account, would COMMON_APPDATA be different than ALLUSERSPROFILE? I assume not, but wanted to check

**Comment by joshcooper** [comment]
👍  to reverting the commits above to avoid confusion

File: ext/windows/service/daemon.rb (hunk 1 above)
Comment (GabrielNagy) [reply ext/windows/service/daemon.rb:20]: I set up a domain controller and client, set the puppet service to run as a domain account and confirmed that COMMON_APPDATA and ALLUSERSPROFILE are the same.

**Comment by GabrielNagy** [comment]
I reverted the `rdoc` commit, the other one is not applicable anymore since `builder.rb` was removed in https://github.com/puppetlabs/puppet/commit/8e53afb395e9810c00982106fa22bc72b3a20665.

**Review by joshcooper — COMMENTED** [review]
LGTM. There's a conflict that needs resolving. Also could you run with pr tests on 32 and 64 bit Windows?

**Comment by GabrielNagy** [comment]
jenkins please test this on windows10ent-64,windows10ent-32,windows2012r2-64,windows2016-64,windows2019-64,windows2012-64,windows2012r2-64,windows2012r2_wmf5-64,windows2012r2_core-64,windows7-64,windows81-64,windows10ent-32,windows10ent-32,windows2016-64,windows2016_core-64,windows2019_core-64,windows2019_ja-64a

**Review by luchihoratiu — APPROVED** [review]

**Review by joshcooper — APPROVED** [review]

**Pull request merged**